File: skellig-test-step-processing-db/src/test/java/org/skellig/teststep/processor/db/DatabaseRequestExecutorFactoryTest.java
package org.skellig.teststep.processor.db;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.skellig.teststep.processing.exception.TestStepProcessingException;
import org.skellig.teststep.processor.db.model.DatabaseRequest;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class DatabaseRequestExecutorFactoryTest {

    private DatabaseRequestExecutorFactory factory;
    private DatabaseRequestExecutor selectExecutor;
    private DatabaseRequestExecutor insertExecutor;

    @BeforeEach
    void setUp() {
        selectExecutor = mock(DatabaseRequestExecutor.class);
        insertExecutor = mock(DatabaseRequestExecutor.class);

        factory = new DatabaseRequestExecutorFactory(selectExecutor, insertExecutor);
    }

    @Test
    void testGetRequestExecutorForSelectCommand() {
        DatabaseRequest databaseRequest = mock(DatabaseRequest.class);
        when(databaseRequest.getCommand()).thenReturn("select");

        DatabaseRequestExecutor result = factory.get(databaseRequest);

        assertEquals(selectExecutor, result);
    }

    @Test
    void testGetRequestExecutorForInsertCommand() {
        DatabaseRequest databaseRequest = mock(DatabaseRequest.class);
        when(databaseRequest.getCommand()).thenReturn("insert");

        DatabaseRequestExecutor result = factory.get(databaseRequest);

        assertEquals(insertExecutor, result);
    }

    @Test
    void testGetRequestExecutorForInsertQuery() {
        DatabaseRequest databaseRequest = mock(DatabaseRequest.class);
        when(databaseRequest.getQuery()).thenReturn(" insert into V(select, gg, bb)");

        DatabaseRequestExecutor result = factory.get(databaseRequest);

        assertEquals(insertExecutor, result);
    }

    @Test
    void testGetRequestExecutorForSelectQuery() {
        DatabaseRequest databaseRequest = mock(DatabaseRequest.class);
        when(databaseRequest.getQuery()).thenReturn(" select * from V when a = insert");

        DatabaseRequestExecutor result = factory.get(databaseRequest);

        assertEquals(selectExecutor, result);
    }

    @Test
    void testGetRequestExecutorForUndefinedQuery() {
        DatabaseRequest databaseRequest = mock(DatabaseRequest.class);
        when(databaseRequest.getQuery()).thenReturn("some query");

        TestStepProcessingException ex =
                assertThrows(TestStepProcessingException.class, () -> factory.get(databaseRequest));

        assertEquals("No database query executors found for query: 'some query'." +
                " Supported types of queries: [select, insert]", ex.getMessage());
    }

    @Test
    void testGetRequestExecutorForUndefinedCommand() {
        DatabaseRequest databaseRequest = mock(DatabaseRequest.class);
        when(databaseRequest.getCommand()).thenReturn("wrong command");

        TestStepProcessingException ex =
                assertThrows(TestStepProcessingException.class, () -> factory.get(databaseRequest));

        assertEquals("No database query executors found for command: 'wrong command'." +
                " Supported commands: [select, insert]", ex.getMessage());
    }
}

File: skellig-test-step-processing-db/src/test/kotlin/org/skellig/teststep/processor/db/DatabaseRequestExecutorFactoryTest.kt
package org.skellig.teststep.processor.db

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.skellig.teststep.processing.exception.TestStepProcessingException
import org.skellig.teststep.processor.db.model.DatabaseRequest

class DatabaseRequestExecutorFactoryTest {

    private var factory: DatabaseRequestExecutorFactory? = null
    private var selectExecutor = Mockito.mock(DatabaseRequestExecutor::class.java)
    private var insertExecutor = Mockito.mock(DatabaseRequestExecutor::class.java)

    @BeforeEach
    fun setUp() {
        factory = DatabaseRequestExecutorFactory(selectExecutor, insertExecutor)
    }

    @Test
    fun testGetRequestExecutorForSelectCommand() {
        val databaseRequest = DatabaseRequest(command = "select")

        val result = factory!![databaseRequest]

        Assertions.assertEquals(selectExecutor, result)
    }

    @Test
    fun testGetRequestExecutorForInsertCommand() {
        val databaseRequest = DatabaseRequest(command = "insert")

        val result = factory!![databaseRequest]

        Assertions.assertEquals(insertExecutor, result)
    }

    @Test
    fun testGetRequestExecutorForInsertQuery() {
        val databaseRequest = DatabaseRequest(query = " insert into V(select, gg, bb)")

        val result = factory!![databaseRequest]

        Assertions.assertEquals(insertExecutor, result)
    }

    @Test
    fun testGetRequestExecutorForSelectQuery() {
        val databaseRequest = DatabaseRequest(query = " select * from V when a = insert")

        val result = factory!![databaseRequest]

        Assertions.assertEquals(selectExecutor, result)
    }

    @Test
    fun testGetRequestExecutorForUndefinedQuery() {
        val databaseRequest = DatabaseRequest(query = "some query")

        val ex = Assertions.assertThrows(TestStepProcessingException::class.java) { factory!![databaseRequest] }

        Assertions.assertEquals("No database query executors found for query: 'some query'." +
                " Supported types of queries: [select, insert]", ex.message)
    }

    @Test
    fun testGetRequestExecutorForUndefinedCommand() {
        val databaseRequest = DatabaseRequest(command = "wrong command")

        val ex = Assertions.assertThrows(TestStepProcessingException::class.java) { factory!![databaseRequest] }

        Assertions.assertEquals("No database query executors found for command: 'wrong command'." +
                " Supported commands: [select, insert]", ex.message)
    }
}